stdlib/public/core/StringStorageBridge.swift

```diff
@@ -335,6 +373,464 @@ extension __SharedStringStorage {
     // reference will make the instance non-unique.
     return self
   }
+  
+  @objc(copy)
+  final internal func copy() -> AnyObject {
+    return self
+  }
+  
+  @objc(mutableCopyWithZone:)
+  final internal func mutableCopy(with zone: _SwiftNSZone?) -> AnyObject {
+    return _SwiftNSMutableString(self.asString)
+  }
+  
+  @objc(mutableCopy)
+  final internal func mutableCopy() -> AnyObject {
+    return _SwiftNSMutableString(self.asString)
+  }
+  
+}
+
+@usableFromInline
+@objc internal final class _SwiftNSMutableString :
+  _SwiftNativeNSMutableString, _AbstractStringStorage {
+  
+  // Swift String always wants to have valid contents, so for cases where
+  // mutation of the contents makes them invalid (e.g. deleting half of a
+  // surrogate pair) we have an alternate backing store that's just an
+  // Array of UTF16, which we largely leave up to NSString to interpret
+  private enum Contents {
+    case native(String)
+    case utf16([UTF16.CodeUnit])
+    
+    var asString: String {
+      @_effects(readonly) @inline(__always) get {
+        switch self {
+        case .native(let str):
+          return str
+        case .utf16(let arr):
+          return String(decoding: arr, as: UTF16.self)
+        }
+      }
+    }
+    
+    var start: UnsafePointer<UInt8>? {
+      @_effects(readonly) @inline(__always) get {
+        switch self {
+        case .native(let str):
+          let guts = str._guts
+          guard !guts.isSmall else { return nil }
+          return guts._object.fastUTF8.baseAddress!
+        default:
+          return nil
+        }
+      }
+    }
+    
+    var isASCII: Bool {
+      @_effects(readonly) @inline(__always) get {
+        switch self {
+        case .native(let str):
+          return str._guts.isASCII
+        default:
+          return false
+        }
+      }
+    }
+    
+    var UTF16Length: Int {
+      @_effects(readonly) @inline(__always) get {
+        switch self {
+        case .native(let str):
+          return str.utf16.count // UTF16View special-cases ASCII for us.
+        case .utf16(let arr):
+          return arr.count
+        }
+      }
+    }
+    
+    @_effects(readonly)
+    func character(at offset: Int) -> UInt16 {
+      switch self {
+      case .native(let str):
+        return str.utf16[str._toUTF16Index(offset)]
+      case .utf16(let arr):
+        return arr[offset]
+      }
+    }
+    
+    func _validUTF16(_ data: UnsafeBufferPointer<UTF16.CodeUnit>) -> Bool {
```

Comment:
Probably not an instance method. Either make it static, or would it make sense to be in UnicodeHelpers.swift?

Reply:
ah, yeah, probably UnicodeHelpers.swift material